MediaBrowser.Controller/SyncPlay/Queue/PlayQueueManager.cs

```diff
@@ -0,0 +1,596 @@
+        /// </summary>
+        /// <param name="items">The new items of the playlist.</param>
+        public void SetPlaylist(Guid[] items)
+        {
+            SortedPlaylist = CreateQueueItemsFromArray(items);
+            PlayingItemIndex = -1;
+            ShuffleMode = GroupShuffleMode.Sorted;
+            RepeatMode = GroupRepeatMode.RepeatNone;
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Appends new items to the playlist. The specified order is mantained for the sorted playlist, whereas items get shuffled for the shuffled playlist.
+        /// </summary>
+        /// <param name="items">The items to add to the playlist.</param>
+        public void Queue(Guid[] items)
+        {
+            var newItems = CreateQueueItemsFromArray(items);
+            SortedPlaylist.AddRange(newItems);
+
+            if (ShuffleMode.Equals(GroupShuffleMode.Shuffle))
+            {
+                newItems.Shuffle();
+                ShuffledPlaylist.AddRange(newItems);
+            }
+
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Shuffles the playlist. Shuffle mode is changed.
+        /// </summary>
+        public void ShufflePlaylist()
+        {
+            if (SortedPlaylist.Count() == 0)
+            {
+                return;
+            }
+
+            if (PlayingItemIndex < 0) {
+                ShuffledPlaylist = SortedPlaylist.ToList();
+                ShuffledPlaylist.Shuffle();
+            }
+            else
+            {
+                var playingItem = SortedPlaylist[PlayingItemIndex];
+                ShuffledPlaylist = SortedPlaylist.ToList();
+                ShuffledPlaylist.RemoveAt(PlayingItemIndex);
+                ShuffledPlaylist.Shuffle();
+                ShuffledPlaylist = ShuffledPlaylist.Prepend(playingItem).ToList();
+                PlayingItemIndex = 0;
+            }
+
+            ShuffleMode = GroupShuffleMode.Shuffle;
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Resets the playlist to sorted mode. Shuffle mode is changed.
+        /// </summary>
+        public void SortShuffledPlaylist()
+        {
+            if (PlayingItemIndex >= 0)
+            {
+                var playingItem = ShuffledPlaylist[PlayingItemIndex];
+                PlayingItemIndex = SortedPlaylist.IndexOf(playingItem);
+            }
+
+            ShuffledPlaylist.Clear();
+
+            ShuffleMode = GroupShuffleMode.Sorted;
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Clears the playlist.
+        /// </summary>
+        /// <param name="clearPlayingItem">Whether to remove the playing item as well.</param>
+        public void ClearPlaylist(bool clearPlayingItem)
+        {
+            var playingItem = SortedPlaylist[PlayingItemIndex];
```

Comment:
It is assumed that the sorted playlist is being used - shouldn't this check to see and get the index of the playing item out of the correct playlist - the function is, after all, public.

Reply:
Yep, my bad here when translating.